velox/functions/prestosql/DateTimeFunctions.h

```diff
@@ -1500,4 +1505,36 @@ struct ToISO8601Function {
   }
 };
 
+template <typename T>
+struct AtTimezoneFunction : public TimestampWithTimezoneSupport<T> {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  std::optional<int64_t> targetTimezoneID_;
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& config,
+      const arg_type<TimestampWithTimezone>* /*tsWithTz*/,
+      const arg_type<Varchar>* timezone) {
+    if (timezone != nullptr) {
+      targetTimezoneID_ = util::getTimeZoneID((std::string_view)(timezone));
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
```

Comment:
@svm1: @mbasmanova just brought up a good point that the Presto java implementation seems a bit more involved (com/facebook/presto/operator/scalar/DateTimeFunctions.java) 

could you check why that is the case? Looks like the might be something not correctly handled in this simpler implementation.

Reply:
@pedroerp posting Presto implementation here for reference:

```
@ScalarFunction(value = "at_timezone", visibility = HIDDEN)
    @LiteralParameters("x")
    @SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE)
    public static long timestampAtTimeZone(@SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE) long timestampWithTimeZone, @SqlType("varchar(x)") Slice zoneId)
    {
        try {
            return packDateTimeWithZone(unpackMillisUtc(timestampWithTimeZone), zoneId.toStringUtf8());
        }
        catch (NotSupportedException | TimeZoneNotSupportedException e) {
            throw new PrestoException(NOT_SUPPORTED, e.getMessage(), e);
        }
        catch (IllegalArgumentException e) {
            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e.getMessage(), e);
        }
    }
```

`packDateTimeWithZone()` as follows:
```
public static long packDateTimeWithZone(long millisUtc, TimeZoneKey timeZoneKey)
    {
        requireNonNull(timeZoneKey, "timeZoneKey is null");
        return pack(millisUtc, timeZoneKey.getKey());
    }
```

Seems to be doing pretty much the same thing, bitpacking the millisecond and timezone ID values. 

As for the exception handling - the equivalent of the `TimeZoneNotSupportedException` is handled implicitly here by the call to `getTimeZoneID()` which would throw a similar error if given an invalid timezone.  

The `IllegalArgumentException` should implicitly be handled by the coordinator if I'm not mistaken - as in any attempt to call the function on invalid/mismatched input should be caught before coming to this level?

Now it is interesting that all the Presto `at_timezone()` implementations take in either `time with time zone` or `timestamp with time zone` - none that take in a mere `timestamp`, as I would expect. Perhaps the values are casted prior to hitting the function, paired with session time zone?

Reply:
@svm1 I wonder if you know what this code is for:

```
    private static long timeAtTimeZone(SqlFunctionProperties properties, long timeWithTimeZone, TimeZoneKey timeZoneKey)
    {
        DateTimeZone sourceTimeZone = getDateTimeZone(unpackZoneKey(timeWithTimeZone));
        DateTimeZone targetTimeZone = getDateTimeZone(timeZoneKey);
        long millis = unpackMillisUtc(timeWithTimeZone);

        // STEP 1. Calculate source UTC millis in session start
        millis += valueToSessionTimeZoneOffsetDiff(properties.getSessionStartTime(), sourceTimeZone);

        // STEP 2. Calculate target UTC millis in 1970
        millis -= valueToSessionTimeZoneOffsetDiff(properties.getSessionStartTime(), targetTimeZone);

        // STEP 3. Make sure that value + offset is in 0 - 23:59:59.999
        long localMillis = millis + targetTimeZone.getOffset(0);
        // Loops up to 2 times in total
        while (localMillis > TimeUnit.DAYS.toMillis(1)) {
            millis -= TimeUnit.DAYS.toMillis(1);
            localMillis -= TimeUnit.DAYS.toMillis(1);
        }
        while (localMillis < 0) {
            millis += TimeUnit.DAYS.toMillis(1);
            localMillis += TimeUnit.DAYS.toMillis(1);
        }

        try {
            return packDateTimeWithZone(millis, timeZoneKey);
        }
        catch (NotSupportedException | TimeZoneNotSupportedException e) {
            throw new PrestoException(NOT_SUPPORTED, e.getMessage(), e);
        }
        catch (IllegalArgumentException e) {
            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e.getMessage(), e);
        }
    }
```

Reply:
@mbasmanova @pedroerp To be entirely honest, this block is throwing me for a loop. 

The function `valueToSessionTimeZoneOffsetDiff()`, which is calculating a given time zone's difference in offsets between the first moment of the epoch (1970-01-01) and the session start time, has some comments explaining its purpose and the greater logic of this process, but I'm still not clear on what exactly is being achieved. 

`millis` is already representing a point in UTC, so I don't see why the function doesn't simply begin from "Step 3" where the target time zone's offset is added to it, to convert to the target time zone. 

But I do see this seems specific to Presto's `time with time zone` type and the limitations around it - namely that lacking a date component makes it more ambiguous to pin down the time zone the input value should be associated with, seems that all this logic is to circumvent that issue, if I'm not mistaken. I don't think this necessarily applies to our situation in Velox - we don't have a `time with time zone` type at all, and never face this ambiguity with `timestamp with time zone` input which is clearly demarcated by its own "starting" time zone.

Reply:
@svm1 Thank you for taking a look. I was puzzled by this as well. As you pointed out this logic applies to TIME with time zone, not TIMESTAMP with time zone. I missed that part. Thank you for clarifying.

> lacking a date component makes it more ambiguous to pin down the time zone the input value should be associated with, seems that all this logic is to circumvent that issue

Sounds plausible.